frame/staking/src/pallet/impls.rs

```diff
@@ -997,29 +997,35 @@ impl<T: Config> ElectionDataProvider for Pallet<T> {
 	type AccountId = T::AccountId;
 	type BlockNumber = BlockNumberFor<T>;
 	type MaxVotesPerVoter = T::MaxNominations;
+	type MaxElectingVoters = <T::ElectionProvider as ElectionProviderBase>::MaxElectingVoters;
+	type MaxElectableTargets = <T::ElectionProvider as ElectionProviderBase>::MaxElectableTargets;
 
 	fn desired_targets() -> data_provider::Result<u32> {
 		Self::register_weight(T::DbWeight::get().reads(1));
 		Ok(Self::validator_count())
 	}
 
-	fn electing_voters(maybe_max_len: Option<usize>) -> data_provider::Result<Vec<VoterOf<Self>>> {
+	fn electing_voters(
+		maybe_max_len: Option<usize>,
+	) -> data_provider::Result<BoundedVec<VoterOf<Self>, Self::MaxElectingVoters>> {
 		// This can never fail -- if `maybe_max_len` is `Some(_)` we handle it.
 		let voters = Self::get_npos_voters(maybe_max_len);
 		debug_assert!(maybe_max_len.map_or(true, |max| voters.len() <= max));
 
-		Ok(voters)
+		Ok(BoundedVec::truncate_from(voters))
 	}
 
-	fn electable_targets(maybe_max_len: Option<usize>) -> data_provider::Result<Vec<T::AccountId>> {
+	fn electable_targets(
+		maybe_max_len: Option<usize>,
+	) -> data_provider::Result<BoundedVec<T::AccountId, Self::MaxElectableTargets>> {
 		let target_count = T::TargetList::count();
 
 		// We can't handle this case yet -- return an error.
 		if maybe_max_len.map_or(false, |max_len| target_count > max_len as u32) {
 			return Err("Target snapshot too big")
 		}
 
-		Ok(Self::get_npos_targets(None))
```

Comment:
If we do have to truncate, may be better to pass the bounds and leave the responsibility to `get_npos_targets` to figure out how to truncate. Since the order is sorted (almost), its not too bad to do this but from this outer method's pov, appears like an unsafe truncate.

Reply:
Good catch, I've refactored get_npos_* to return a bounded vec and so we don't need to truncate here anymore.